fsthttp/cache.go

```diff
@@ -0,0 +1,761 @@
+// body, and the `transform` callback will be invoked to return a new
+// `io.ReadCloser` for the transformed body.
+func (candidateResponse *CandidateResponse) SetBodyTransform(fn func(io.ReadCloser) io.ReadCloser) {
+	candidateResponse.bodyTransform = fn
+}
+
+// BodyTransform returns the current body transformation function.
+func (candidateResponse *CandidateResponse) BodyTransform() func(io.ReadCloser) io.ReadCloser {
+	return candidateResponse.bodyTransform
+}
+
+func (candidateResponse *CandidateResponse) getSuggestedCacheWriteOptions() (*cacheWriteOptions, error) {
+	if candidateResponse.suggestedCacheWriteOptions == nil {
+		opts, err := candidateResponse.buildFreshSuggestedCacheWriteOptions()
+		if err != nil {
+			return nil, err
+		}
+		candidateResponse.suggestedCacheWriteOptions = opts
+	}
+
+	return candidateResponse.suggestedCacheWriteOptions, nil
+}
+
+func (candidateResponse *CandidateResponse) buildFreshSuggestedCacheWriteOptions() (*cacheWriteOptions, error) {
+	opts, err := httpCacheGetSuggestedCacheWriteOptions(candidateResponse.cacheHandle, candidateResponse.abiResp)
+	if err != nil {
+		return nil, err
+	}
+
+	keys := opts.surrogate
+	if extra := candidateResponse.extraSurrogateKeys; extra != "" {
+		if keys != "" {
+			keys += " "
+		}
+		keys += extra
+	}
+
+	const fastlyHeaderSurrogateKey = "surrogate-key"
+
+	vals := candidateResponse.abiResp.GetHeaderValues(fastlyHeaderSurrogateKey, RequestLimits.maxHeaderNameLen)
+	for vals.Next() {
+		if keys != "" {
+			keys += " "
+		}
+		keys += string(vals.Bytes())
+	}
+	if err := vals.Err(); err != nil {
+		return nil, fmt.Errorf("get header values: %w", err)
+	}
+	opts.surrogate = keys
+
+	return opts, nil
+}
+
+func httpCacheGetSuggestedCacheWriteOptions(cacheHandle *fastly.HTTPCacheHandle, resp *fastly.HTTPResponse) (*cacheWriteOptions, error) {
+	var opts cacheWriteOptions
+	opts.abiOpts.FillConfigMask()
+
+	newABIOpts, err := fastly.HTTPCacheGetSuggestedCacheOptions(cacheHandle, resp, &opts.abiOpts)
+	if err != nil {
+		return nil, fmt.Errorf("get suggested cache options: %w", err)
+	}
+
+	newOpts := &cacheWriteOptions{abiOpts: *newABIOpts}
+	newOpts.loadFromABI()
+
+	return newOpts, nil
+}
+
+func (candidateResponse *CandidateResponse) finalizeOptions() (fastly.HTTPCacheStorageAction, *cacheWriteOptions, error) {
+	var storageAction = candidateResponse.suggestedStorageAction
+	if candidateResponse.useStorageAction {
+		storageAction = candidateResponse.overrideStorageAction
+	}
+
+	suggestedCacheWriteOptions := candidateResponse.suggestedCacheWriteOptions
+	if suggestedCacheWriteOptions == nil {
+		var err error
+		suggestedCacheWriteOptions, err = candidateResponse.buildFreshSuggestedCacheWriteOptions()
+		if err != nil {
+			return 0, nil, err
+		}
+	}
+
+	var opts cacheWriteOptions
+
+	if candidateResponse.useTTL {
+		opts.maxAge = candidateResponse.overrideTTL - suggestedCacheWriteOptions.age
```

Comment:
Why subtract `age` here?

Reply:
https://docs.rs/fastly/latest/src/fastly/http/candidate_response.rs.html#240

Reply:
Ok. Reading that rust, the names don't directly translate, so I'll have to take a closer look at a more opportune time.

Reply:
The translation looks faithful now that I've been able to compare directly.